### PR TITLE
Add `Functor`, `Foldable`, `Traversable` instances

### PR DESCRIPTION
**Pull request by paulyoung**
My original motivation for this was addressed with https://github.com/paulyoung/purescript-corefn/pull/21 but I think this is probably still useful to have.

### CODE REVIEW

File: src/CoreFn/Expr.purs [hunk 1]
```diff
@@ -57,6 +59,22 @@ derive instance eqLiteral :: Eq a => Eq (Literal a)
...
+  foldMap _ _ = mempty
+
+  foldl f = foldlDefault f
+  foldr f = foldrDefault f
+
+instance functorLiteral :: Functor Literal where
+  map _ (NumericLiteral x) = NumericLiteral x
+  map _ (StringLiteral x) = StringLiteral x
+  map _ (CharLiteral x) = CharLiteral x
+  map _ (BooleanLiteral x) = BooleanLiteral x
```
Comment (paulyoung) [inline src/CoreFn/Expr.purs]: I tried doing `map _ x = x` here but got a type error.

File: src/CoreFn/Expr.purs [hunk 2]
```diff
@@ -65,6 +83,16 @@ instance showLiteral :: Show a => Show (Literal a) where
   show (ArrayLiteral a) = "(ArrayLiteral " <> show a <> ")"
   show (ObjectLiteral o) = "(ObjectLiteral" <> show o <> ")"
 
+instance traversableLiteral :: Traversable Literal where
+  traverse _ (NumericLiteral x) = pure (NumericLiteral x)
+  traverse _ (StringLiteral x) = pure (StringLiteral x)
+  traverse _ (CharLiteral x) = pure (CharLiteral x)
+  traverse _ (BooleanLiteral x) = pure (BooleanLiteral x)
+  traverse f (ArrayLiteral x) = ArrayLiteral <$> traverse f x
+  traverse f (ObjectLiteral x) = ObjectLiteral <$> traverse (traverse f) x
```
Comment (paulyoung) [inline src/CoreFn/Expr.purs]: Same here with `traverse _ x = pure x`.

File: src/CoreFn/Expr.purs [hunk 3]
```diff
@@ -65,6 +83,16 @@ instance showLiteral :: Show a => Show (Literal a) where
   show (ArrayLiteral a) = "(ArrayLiteral " <> show a <> ")"
   show (ObjectLiteral o) = "(ObjectLiteral" <> show o <> ")"
 
+instance traversableLiteral :: Traversable Literal where
+  traverse _ (NumericLiteral x) = pure (NumericLiteral x)
+  traverse _ (StringLiteral x) = pure (StringLiteral x)
+  traverse _ (CharLiteral x) = pure (CharLiteral x)
+  traverse _ (BooleanLiteral x) = pure (BooleanLiteral x)
```
Comment (paulyoung) [inline src/CoreFn/Expr.purs]: Same here with `traverse _ x = pure x`.

File: src/CoreFn/Expr.purs (hunk 1 above)
Comment (LiamGoodacre) [inline src/CoreFn/Expr.purs]: You should now be able to use `derive instance functorLiteral :: Functor Literal` :)

Comment (paulyoung) [inline src/CoreFn/Expr.purs]: I know, its awesome! Thanks!

**Comment by paulyoung** [comment]
I think it's probably also worth including `Bind` in this pull request.

**Comment by paulyoung** [comment]
Closing in favor of #42.

**Pull request closed without merging**